Shopping cart: Add CartMessages to CalypsoShoppingCartProvider

sirbrillig opened the pull request:
#### Changes proposed in this Pull Request

This renders the `CartMessages` component everywhere that `CalypsoShoppingCartProvider` is rendered, which will make sure that any success or failure messages returned by the shopping cart endpoint will be displayed.

Fixes https://github.com/Automattic/wp-calypso/issues/48654

#### Testing instructions

1. Starting at URL: http://calypso.localhost:3000/domains/add/example.com
2. Apply D54847-code to your sandbox and sandbox the API. This will break adding G Suite products.
3. Add a domain to your cart.
4. At the upsell, add G Suite to your cart.
5. Verify that you end up in checkout, but that the G Suite product is not in the cart and that you see an appropriate error message.

<img width="1106" alt="Screen Shot 2021-01-04 at 6 23 39 PM" src="https://user-images.githubusercontent.com/2036909/103590057-e652c600-4eba-11eb-9b6e-e48169330ef3.png">

sirbrillig committed:
Remove CartMessages from PopoverCart

sirbrillig committed:
Remove CartMessages from CompositeCheckout

sirbrillig committed:
Get selectedSiteSlug from Redux in CartMessages

sirbrillig committed:
Add CartMessages to CalypsoShoppingCartProvider

stephanethomas commented:
I couldn't make that fix work but I'm not sure if that's because this pull request should be rebased now that #48202 has been merged, or if that's because of some of my own changes (I tested it on calypso.live with D54813-code). It's already late here, I'll have another look to my patch tomorrow to see if it is the culprit but I thought I should mention my lack of success in case that helps 🙂. For the records, here are the two use cases that I've tested:

Use case #1

1. Navigate to the `Domain Settings` of a domain already purchased
2. Add Google Workspace with a valid user but for an account no longer with us
3. `modify_cart_after_adding_product()` removes Google Workspace from the cart and adds an error to it
4. `Plans` page is displayed with no error message

Use case #2

1. Add a domain to the cart
2. Add Google Workspace with an invalid user
3. `modify_cart_after_adding_product()` removes Google Workspace from the cart and adds an error to it
4. `Checkout` is displayed with only the domain (good) but no error message (bad)

sirbrillig commented:
Thanks for the reminder, and the testing instructions! I rebased this PR and I'll test it again myself later today. It may be that there's a case this doesn't cover.

pottedmeat reviewed (approved):
Code looks good

pottedmeat commented on client/my-sites/checkout/cart/popover-cart.jsx:
```diff
@@ -16,7 +16,6 @@ import { withShoppingCart } from '@automattic/shopping-cart';
  */
 import CartBody from './cart-body';
 import CartBodyLoadingPlaceholder from './cart-body/loading-placeholder';
-import CartMessages from './cart-messages';
```
Is `<PopoverCart>` within `<CalypsoShoppingCartProvider>`? It uses `withShoppingCart` and I'm still looking through its position within the component tree.

sirbrillig replied:
Yup! See https://github.com/Automattic/wp-calypso/pull/47605

pottedmeat replied:
I see it wrapped as part of `<PlansNavigation>` but I'm having a hard time finding it through `<HeaderCart>` as I'm not really sure how the `<Main>` component is used where it eventually ends up.

nbloomf commented:
After following the test instructions, the GSuite sub is removed from the cart but I don't see a notification.

sirbrillig committed:
Update GSuiteUpgrade to use useShoppingCart for everything

sirbrillig committed:
Convert CartMessages to a functional component

sirbrillig commented:
Wow, weird. You're right... it looks like when I rebased this, I totally failed to actually rebase it onto the changes from #48202. (In addition, those changes were incomplete; `cart` had not been pulled from `useShoppingCart`, although I don't think that actually caused any issues since it only would have broken a fallback redirect.)

I rebased it again and now I think it works. My apologies for whatever I did before!

stephanethomas commented:
I tested it again but I'm still not getting any error message with use case #1 (when I land on the `Plans` page). Use case #2 works fine now.

sirbrillig commented:
Oh! Sorry. I think that https://github.com/Automattic/wp-calypso/pull/48687 should take care of your first use case. Perhaps we should merge that one first so that we can test that properly here, although the two PRs don't technically overlap.

stephanethomas commented:
No worries!

>  I think that #48687 should take care of your first use case.

I've tested that by checking out #48658 and cherry picking #48687. I could indeed see the error message while the checkout page was loading (and showing placeholders). However, this error message disappeared two seconds later when I was redirected to the `Plans` page.

> Perhaps we should merge that one first so that we can test that properly here, although the two PRs don't technically overlap.

I'm fine with that 👍.

sirbrillig commented:
> However, this error message disappeared two seconds later when I was redirected to the Plans page.

Huh... the notice is supposed to be persistent, meaning it requires user interaction to dismiss, but maybe something else is clearing the notice. I'll have another look.

stephanethomas commented:
> the notice is supposed to be persistent, meaning it requires user interaction to dismiss, but maybe something else is clearing the notice.

To be more precise, the error message for G Suite is replaced by another one about the Live Stripe.js integration. However, the latter disappears when I'm redirected to the `Plans` page.

sirbrillig commented:
> To be more precise, the error message for G Suite is replaced by another one about the Live Stripe.js integration. However, the latter disappears when I'm redirected to the Plans page.

Ah! That's probably what's happening. IIRC the notices library only lets you display one notice at a time (weird, I know) and the "Live Stripe.js" notification is not persistent, so when it's displayed, it removes the G Suite notice, and then that notice is removed by the page redirect.

Probably what we should do is change the notices library to allow persistent notices to remain even if another notice is added, but that might be out of the scope of this PR. The "Live Stripe.js" notification should not occur in production (it only happens when you have calypso.localhost using the production API) so the bug you're seeing should be an edge case for now.

sirbrillig commented:
Ok, given the above conversation, I think all the use-cases should be covered. I'm going to merge this but we can double check everything once it's in production and make sure there's no cases we didn't catch. In any case, this PR shouldn't make things any worse and should improve the situation for most error messages.
